### PR TITLE
Add g-key support for use in global shortcuts

### LINKED ISSUE #1721: [Feature Request] Logitech G-Keys
I'd like to write a PR for this! Would you accept adding this, esp. since the LCD plugin is already there? Who do I have to contact to get help, especially with modifying the windows build process (releng?) to include logitech's library?

FYI, for some reason, the current version of the g-key library available on [logitech's site](http://gaming.logitech.com/developers) ([GkeySDK_8.57.148.zip](http://www.logitech.com/pub/techsupport/SDK/GkeySDK_8.57.148.zip)) doesn't seem to work with my g710 with version 8.50.281 of logitech's desktop software (their current version is 8.58.183). However, it does work with the older [1.02.004 g-key library](https://github.com/noorus/nil/tree/master/external/LogitechGkeySDK_1.02.004) . I'll probably test upgrading my logitech software to see if it works with the new version then.

I'm currently working on [a tray app to talk to a teamspeak plugin](https://github.com/forivall/niftykb) to enable push to talk for keys that only directinput seems to pick up, where I basically ripped out mumble's keybinding code (so I've got a decent understanding of how do implement this). I also wanted to use it to allow mumble to detect g-key events, but I just realized that your plugin system doesn't do that kind of thing, so I'm submitting this PR. Alternatively, adding more features to the plugin system could be added, but I don't have the expertise needed for that.

### PR DESCRIPTION
**Pull request by forivall**
Fixes #1721

There are still some todos for this before it should be merged:

1. <s>The location of the dll needs to be loaded by looking at the registry, instead of how it is currently hardcoded</s> see checklist.
2. <s>code review plz! I modeled it after CELTCodec.cpp/h.</s> "LGTM"
3. If somebody has a logitech mouse with billions of buttons, it would be nice if they could test this (G600, G300, G400, G700, G9/G9x, MX518, G402, G502, or G602). Note that this is only for the extra buttons, higher than button 5.
4. <s>should the `QList<QVariant>` format for passing around buttons be made uniform, so that the first element is always a signifier of if it's a dinput or a gkey? or should i create a uuid for gkeys, and keep the list length at 2?</s> Create a uuid for gkeys and gmousebuttons

<s>And finally, the instructions to the user, that will need to be placed somewhere on the wiki: This requires Logitech Gaming Software version 8.55+, and they need to set the mumble profile to "Set as Default" or "Set as Persistent" to allow the keys to work while mumble is in the background.</s> Included in pr.

---

- [x] Rename files and class
- [x] Include Spec and doc in .cpp file
- [x] add hidden config option
- [x] Create guid for gkey key and gkey mousebutton
- [x] update reporting to use guid
- [x] query registry for dll location
- [x] update build settings
- [ ] Also add user documentation to the wiki

### CODE REVIEW

File: src/mumble/GlobalShortcut_win.cpp [hunk 1]
```diff
@@ -557,12 +566,53 @@ void GlobalShortcutWin::timeTicked() {
 			handleButton(ql, rgdod[j].dwData & 0x80);
 		}
 	}
+#ifdef USE_GKEYS
+	if (gkey->isValid()) {
+		for (int button = GKEY_MIN_MOUSE_BUTTON; button <= GKEY_MAX_MOUSE_BUTTON; button++) {
```
Comment (mkrautz) [inline src/mumble/GlobalShortcut_win.cpp]: Indentation.

File: src/mumble/GlobalShortcut_win.cpp [hunk 2]
```diff
@@ -557,12 +566,53 @@ void GlobalShortcutWin::timeTicked() {
...
+            QList<QVariant> ql;
+            ql << GKEY_BUTTON_MOUSE;
+            ql << button;
+            ql << 0; // 3 entries in the list means gkey; add '0' as a dummy
+            handleButton(ql, gkey->isMouseButtonPressed(button));
+		}
+		for (int mode = GKEY_MIN_KEYBOARD_MODE; mode <= GKEY_MAX_KEYBOARD_MODE; mode++) {
```
Comment (mkrautz) [inline src/mumble/GlobalShortcut_win.cpp]: Indentation.

File: src/mumble/GlobalShortcut_win.cpp (hunk 1 above)
Comment (forivall) [inline src/mumble/GlobalShortcut_win.cpp]: Wierd, my editor switched to spaces for those lines. Fixed.

**Comment by mkrautz** [comment]
I'd like if this was using a GUID for G-keys instead. Fits the GlobalShortcut_win architecture better.

I'd like it if you added a comment at the top of Gkeys.cpp saying: "The code interfacing with the Logitech G-Keys DLL was implemented using the following spec:\n\n<the spec I wrote>"

Let's rename the file from Gkeys.{cpp,h} to GKeys.{cpp,h}. Very nitpicky.

I think we might as well enable G-keys support by default.
You can easily do that by having a !CONFIG(no-gkeys) { CONFIG *= gkeys } (with correct indentation) just above your CONFIG(gkeys) section.

Code LGTM....  If you implement my spec, registry querying and GUIDs for G-Ke-Keyboard and G-Key-Mouse, I think it's ready to go in.

**Comment by mkrautz** [comment]
I guess also means I prefer the class to be called GKeyLibrary. :-)

**Comment by mkrautz** [comment]
Oh, and the !CONFIG(no-gkeys) would probably need to go after CONFIG(gkeys)...

**Comment by mkrautz** [comment]
Also, I would like to ask you to include a config option (no UI, just a config option) for disabling G-keys suport.

You can base it on https://github.com/mumble-voip/mumble/commit/18c359f82a758ae68886fb4b8b270916032cec9e

I think it should just be called "shortcut/gkey" or something. I don't think we need to put it under "shortcut/win/gkey" since it is not OS-vendor specific. We could grow an implementation for other OSes in the future, so "shortcut/gkey" seems fine.

It should be enabled by default.

In GlobalShortcut_win.cpp, you can then just do (g.s.bEnableGKey) { gkeys = new GKeyLibrary(); }, and gate the polling inside the ifdefs with a similar check.

This allows us to have users apply a registry hotfix to disable G-Key support in case it causes performance problems, hangs or crashes in the future.

**Comment by forivall** [comment]
Awesome! should I just create a guid with `uuidgen`, or is there some other way (hashing something?). 

Should I include the full text of the spec in the comment, or just a link?

I would anticipate that the only other os that this would be needed for is OSX. ([Since LInux users have to use custom drivers :sweat:](https://github.com/Wattos/logitech-g710-linux-driver))

I'll do all of this sometime tomorrow (PDT). It's late.

**Comment by mkrautz** [comment]
Just create a v4 GUID somewhere. I use python: import uuid; uuid.uuid4().

Spec: just include a line stating you implemented it using the spec below, and include it verbatim inside the source file.

**Comment by mkrautz** [comment]
You might also want to add a small doc snipped inside the .cpp file describing how to setup Logitech Gaming Software to interact with Mumble.

You are right that it probably should live in the Wiki for end-users to see, but I think it would be nice to have in the code as well.

**Comment by forivall** [comment]
And done!

File: src/mumble/GKeys.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+	bool isMouseButtonPressed(int button);
+	bool isKeyboardGkeyPressed(int key, int mode);
+	QString getMouseButtonString(int button);
+	QString getKeyboardGkeyString(int key, int mode);
+
+protected:
+	QLibrary qlLogiGkey;
+	bool bValid;
+
+	bool (*LogiGkeyInit)(void *);
```
Comment (mkrautz) [inline src/mumble/GKeys.h]: BOOL as per the spec.

File: src/mumble/GKeys.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	void (*LogiGkeyShutdown)();
+	bool (*LogiGkeyIsMouseButtonPressed)(int button);
```
Comment (mkrautz) [inline src/mumble/GKeys.h]: BOOL as per the spec.

File: src/mumble/GKeys.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	bool (*LogiGkeyIsKeyboardGkeyPressed)(int key, int mode);
```
Comment (mkrautz) [inline src/mumble/GKeys.h]: BOOL as per the spec.

File: src/mumble/GKeys.cpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+		// XXX: Should the unicode variants be used instead?
+		LONG err = RegQueryValueExA(key, "", NULL, &type, reinterpret_cast<LPBYTE>(libLocation), &len);
+		if (err == ERROR_SUCCESS && type == REG_SZ) {
+			qWarning("GKeyLibrary: Found ServerBinary with libLocation = \"%s\", len = %d", libLocation, len);
+			alternatives << QString::fromLatin1(reinterpret_cast<const char *>(libLocation), len);
+		} else {
+			qWarning("GKeyLibrary: Error looking up ServerBinary (Error: 0x%x, Type: 0x%x, len: %d)", err, type, len);
+		}
+	} else {
+		qWarning("GKeyLibrary: Could not open lib location regkey (Error: 0x%x)", errOpen);
```
Comment (mkrautz) [inline src/mumble/GKeys.cpp]: Probably drop this warning. It'll be a lot of noise for people without LGS.

**Comment by mkrautz** [comment]
Looks pretty good now!

I know I suggested GKeys.cpp and GKeys.h, but would it be more correct for it to be GKey.cpp and GKey.h? And USE_GKEY? It fits better with the class name, DLL function names and so on. If it's not too burdensome, I think we should change it to just be GKey all around.

File: src/mumble/GKeys.h [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+   PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+   LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#ifndef MUMBLE_MUMBLE_GKEY_H
+#define MUMBLE_MUMBLE_GKEY_H
+
+// #include <stdint.h>
```
Comment (mkrautz) [inline src/mumble/GKeys.h]: Drop if not needed?

File: src/mumble/GKeys.cpp [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+{
+	QStringList alternatives;
+
+	HKEY key = NULL;
+	DWORD type = 0;
+	LPSTR libLocation[255];
+	DWORD len = 255;
+	LONG errOpen = RegOpenKeyExA(GKEY_LOGITECH_DLL_REG_HKEY, GKEY_LOGITECH_DLL_REG_PATH, NULL, KEY_READ, &key);
+	if (errOpen == ERROR_SUCCESS) {
+		// XXX: Should the unicode variants be used instead?
```
Comment (mkrautz) [inline src/mumble/GKeys.cpp]: Yes, we simply use RegOpenKeyEx, RegQueryValueEx, etc. in most places in the rest of the codebase.

**Comment by forivall** [comment]
Yeah, I'm fine with renaming stuff (except that when I changed "USE_GKEYS" to "USE_GKEY", I had do do a clean build cuz the pch had the old define. :anguished:) I also changed the qmake flag from `gkeys` to `gkey`, for more consistency. All references now use "GKey" instead of "GKeys", except for one place in the doc comment

**Comment by mkrautz** [comment]
Looks good!  One thing: I can't land this with the current license text.

You've changed "All rights reserved" to "Some rights reserved", which I myself don't particularly mind.

But the "All rights reserved" text is part of the 3-clause BSD license, whether we like it or not. So it ought to stay.

I would have done this myself, but since your modified BSD-license text contains "Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.", I cannot make that change myself, as I would breach the license.

So in order for me to be able to land this, please revert it back to the default 3-clause BSD license used elsewhere in Mumble.

Thanks!

**Comment by forivall** [comment]
Sorry about that. I usually just slap a LICENSE file at the base of my project, and then not really pay attention to the individual file header license text.

**Comment by mkrautz** [comment]
Marking as ready-to-land. LGTM.

I will land once I am able to.

**Comment by mkrautz** [comment]
Landed as 6c096c3148bf610ae85d4cde381d2ff366e3109f.

Very sorry for the long wait. :-(

**Comment by mkrautz** [comment]
Oh, and just to add: Thank you very much, this is pretty cool :-)

**Comment by pyrojoe** [comment]
Any idea when the build with this addition be released?  and If I want to build this myself do I have to use the gkey flag? I want to test this on my G500 and G710+

**Comment by mkrautz** [comment]
I will do a snapshot soon.

**Comment by mkrautz** [comment]
Triggered a snapshot... It's building!

@pyrojoe The gkey CONFIG option is always on, unless disabled with CONFIG-=no-gkey.

**Comment by pyrojoe** [comment]
Thanks @mkrautz I saw the snapshot "mumble-1.3.0~690~ge614a7b~snapshot"

I've tested it and I've got it working on both my keyboard (G710+) and mouse (G500 predecessor to the G502). I had to reinstall LGS for it to work on both my machines.  The downside of this implementation is mumble only reacts to that key while the mumble profile is active (either when mumble is in focus or if you have that profile set as persistent)

This means the mumble profile is the only profile you can use, if you have profiles with bindings for games you can't use them.  So while this sounded like a useful feature on paper it's only useful to those who use one profile across all their applications anyway.

It would also be a viable solution if the G-keys could be assigned in all profiles.  Then I could just have my mouse button set to the G-key for all profiles and not have to worry about it.  But as far as I can tell you can only assign the G-key for programs that support it, and only one program can use a G-key at a time.

To be sure, is my understanding correct @forivall ?  If it is I'm trying to figure out the benefit of this addition.  I didn't have much familiarity with G-Keys so wasn't expecting the use case to be so narrow =/

**Comment by forivall** [comment]
@pyrojoe Yes, unfortunately, that is correct. That's just the way that the LGS works and the way that the G-key API works, and the G-key API is the only way to listen directly to g-keys. I wish that Logitech would allow a way to globally listen to g-keypresses, without requiring a profile. The only other ways to interact with the G-keys are by using the LGS GUI or embedded Lua, which don't lend well to direct integration with mumble. The third solution would be to use the Lua scripting to simulate F13-F18, and fix #987. However, that's a harder bug, and one that I'm not equipped to try to fix :disappointed:. 

This behaviour is [documented in the code (link)](https://github.com/mumble-voip/mumble/commit/6c096c3148bf610ae85d4cde381d2ff366e3109f#diff-f0b8bd1dc5a95211f1673caf2598797fR80), and will be updated in the wiki. I was just waiting for this pr to land before updating the wiki; didn't want to document something that hasn't landed yet.

One thing that you may have missed is that you can also set the mumble profile as the default, so if you have a specific game profile, it will still activate. Personally, I really only use one profile anyways.

**Pull request closed without merging**